lk2nd/lk2nd-device.c

```diff
@@ -275,6 +275,35 @@ static void lk2nd_parse_device_node(const void *fdt)
 #endif
 }
 
+struct lk2nd_keymap* lk2nd_fdt_parse_early_keys(void)
+{
+	int offset, len;
+	const uint32_t *val;
+	struct lk2nd_keymap *map = NULL;
+
+	void *fdt = (void*) lk_boot_args[2];
+	if (!fdt || dev_tree_check_header(fdt))
+		return NULL; // Will be reported later again. Hopefully.
+
+	offset = lk2nd_find_device_offset(fdt);
+	if (offset < 0)
+		return NULL;
+
+	/* FIXME: Assuming that all keys are pulled UP */
+	val = fdt_getprop(fdt, offset, "lk2nd,keys", &len);
+	if (len > 0 && len % 16 == 0) {
+		len /= 16;
+		/* sentinel: key=0 */
+		map = calloc(len + 1, sizeof(struct lk2nd_keymap));
+		for (int i = 0; i < len; i++) {
+			map[i].key = fdt32_to_cpu(val[i*4]);
+			map[i].gpio = fdt32_to_cpu(val[i*4 + 1]);
+			map[i].pull = fdt32_to_cpu(val[i*4 + 2]);
+			map[i].active = fdt32_to_cpu(val[i*4 + 3]);
```

Comment:
Can we put pull + active into one "flags" value maybe? Just to make it a bit more like Linux. Also, the uint32 is probably more than large enough

Reply:
I was trying to avoid moving bits back and forth here and reuse the defines from gpio.h that lk uses. Can probably shift active into the second byte and keep pull as is I guess